File: rf-signal-algorithms/c/pel.cc
/*****************************************************************************
*  Plane Earth Path Loss model for Signal Server by Alex Farrant             *
*  Taken from "Antennas and Propagation for wireless communication systems"  *
*  ISBN 978-0-470-84879-1                                                    *
*  10 August 2016							     *
*  This program is free software; you can redistribute it and/or modify it   *
*  under the terms of the GNU General Public License as published by the     *
*  Free Software Foundation; either version 2 of the License or any later    *
*  version.								     *
*									     *
*  This program is distributed in the hope that it will useful, but WITHOUT  *
*  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or     *
*  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License     *
*  for more details.							     *
*									     */

#include <math.h>

double PlaneEarthLoss(float d, float TxH, float RxH)
{
/*
Plane Earth Loss model 
Frequency: N/A
Distance (km): Any
*/
	// Plane earth loss is independent of frequency.
	double dbloss = 40*log10(d) + 20*log10(TxH) + 20*log10(RxH);
	return dbloss;
}

File: rf-signal-algorithms/src/pel.rs
use crate::Distance;

/// Plane Earth Model
/// Original C implementation: https://github.com/Cloud-RF/Signal-Server/blob/master/models/pel.cc
/// Taken from "Antennas and Propagation for wireless communication systems"  *
/// ISBN 978-0-470-84879-1 (by Alex Farrant)
///
/// Distance (meters) is unbounded.
/// Frequency is not used in this calculation.
/// Transmitter height and receiver height are height AMSL.
pub fn plane_earth_path_loss(tx_height: Distance, rx_height: Distance, distance: Distance) -> f64 {
    let d = distance.as_km();
    let txh = tx_height.as_meters();
    let rxh = rx_height.as_meters();
    (40.0 * d.log10()) + (20.0 * txh.log10()) + (20.0 * rxh.log10())
}
